core/ecschema-metadata/src/SchemaFormatsProvider.ts

```diff
@@ -0,0 +1,122 @@
+import { FormatsProvider } from "./Interfaces";
+import { SchemaItemKey, SchemaKey } from "./SchemaKey";
+import { SchemaItem } from "./Metadata/SchemaItem";
+import { Format } from "./Metadata/Format";
+import { SchemaItemFormatProps } from "./Deserialization/JsonProps";
+import { BeUiEvent } from "@itwin/core-bentley";
+import { KindOfQuantity } from "./Metadata/KindOfQuantity";
+import { getFormatProps } from "./Metadata/OverrideFormat";
+import { UnitSystemKey } from "@itwin/core-quantity";
+
+/**
+ * Provides default formats and kind of quantities from a given SchemaContext or SchemaLocater.
+ * @beta
+ */
+export class SchemaFormatsProvider implements FormatsProvider {
+  private _context: SchemaContext;
+  private _unitSystem: UnitSystemKey;
+  public onFormatsChanged = new BeUiEvent<string[]>();
+  /**
+   *
+   * @param contextOrLocater The SchemaContext or a different ISchemaLocater implementation used to retrieve the schema. The SchemaContext
+   * class implements the ISchemaLocater interface. If the provided locater is not a SchemaContext instance a new SchemaContext will be
+   * created and the locater will be added.
+   * @param _unitExtraData Additional data like alternate display label not found in Units Schema to match with Units; Defaults to empty array.
+   */
+  constructor(contextOrLocater: ISchemaLocater, unitSystem: UnitSystemKey) {
+    if (contextOrLocater instanceof SchemaContext) {
+      this._context = contextOrLocater;
+    } else {
+      this._context = new SchemaContext();
+      this._context.addLocater(contextOrLocater);
+    }
+    this._unitSystem = unitSystem;
+  }
+
+  public get context() { return this._context; }
+  public get unitSystem() { return this._unitSystem; }
+
+  public set unitSystem(unitSystem: UnitSystemKey) {
+    this._unitSystem = unitSystem;
+    this.onFormatsChanged.raiseEvent([this._unitSystem]);
+  }
+
+  private async getKindOfQuantityFormatFromSchema(itemKey: SchemaItemKey, unitSystem: UnitSystemKey): Promise<SchemaItemFormatProps | undefined> {
+    const kindOfQuantity = await this._context.getSchemaItem(itemKey, KindOfQuantity);
+
+    if (!kindOfQuantity) {
+      return undefined;
+    }
+    // Find the first presentation format that matches the provided unit system. Else, use the first entry.
+    if (unitSystem) {
+      const unitSystemGroupNames = getUnitSystemGroupNames(unitSystem);
+      const presentationFormats = kindOfQuantity.presentationFormats;
+      for (const system of unitSystemGroupNames) {
+        for (const format of presentationFormats) {
+          const unit = format.units && format.units[0][0];
+          if (!unit) {
+            continue;
+          }
+          const currentUnitSystem = await unit.unitSystem;
+          if (currentUnitSystem && currentUnitSystem.name.toUpperCase() === system) {
+            return getFormatProps(format);
+          }
+        }
+      }
+      return undefined;
```

Comment:
should we really return here? this is different from [what we have in Presentation packages](https://github.com/iTwin/presentation/blob/master/packages/core-interop/src/core-interop/Formatting.ts), where we:
- fall back to persistence unit format props if their unit system matches requested one,
- fall back to default presentation format even if it doesn't match requested unit system.

Reply:
gotcha, I've removed the return statement here, so we should fall back to the default presentation format (if there is one)

Reply:
do you think the fallback to persistence unit is not necessary anymore?

Reply:
I was a bit hesitant to use it, in particular defaulting to using DefaultRealU and such. What if the KindOfQuantity was targeting non decimal formatTypes and etc. I added that fallback for now, and if we find the scenario above to be a problem we could remove it, or rearrange the order of the fallbacks.